### ISSUE TITLE
Use a modern font stack in aliases used for system font loading (`serif`, `sans-serif`, …)

### ISSUE THREAD

**Issue by Calinou** [godotengine/godot-proposals#5767]
### Describe the project you are working on

The Godot editor :slightly_smiling_face: 

### Describe the problem or limitation you are having in your project

The current system font aliases don't lead to the best possible visual result. Fonts that are selected (such as Arial or Times New Roman on Windows) usually look dated.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Inspired by the [Bootstrap font stack](https://github.com/twbs/bootstrap/blob/0446e22b5aa556701f32015e9678f4ae822d1386/dist/css/bootstrap.css#L52-L53), we can resolve the following aliases this way:

Font alias | Windows | macOS/iOS
-|-|-
`sans-serif` | Segoe UI | San Francisco[^1]
`serif` | Cambria | New York[^1][^2]
`monospace` | Consolas[^3] | SF Mono

[^1]: [Is shipped as a variable font.](https://developer.apple.com/design/human-interface-guidelines/foundations/typography)
[^2]: May not be usable directly according to the macOS documentation. In this case, *Georgia* could be used instead.
[^3]: Cascadia Code exists as a modern replacement, but is only bundled on Windows 11 with Windows Terminal. Consolas is available on Windows 7 and later.

`fantasy` and `cursive` are not changed by this proposal.

Alias resolution on Linux will probably remain handled by fontconfig to avoid issues across various Linux distributions. That said, we could explore trying out specific fonts such as `Roboto`, `Noto Sans`, `Open Sans` then `Liberation Sans`, which should cover most distributions out there with a better-looking font compared to the default `sans-serif`.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

https://github.com/godotengine/godot/blob/c17f17eb98188a7134c85bdbdf0123127c462046/platform/windows/os_windows.cpp#L776-L786

https://github.com/godotengine/godot/blob/c17f17eb98188a7134c85bdbdf0123127c462046/platform/macos/os_macos.mm#L342-L353

https://github.com/godotengine/godot/blob/c17f17eb98188a7134c85bdbdf0123127c462046/platform/ios/os_ios.mm#L340-L350

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Similar to how it's worked around on the web platform, this can be worked around by specifying a list of fonts with appropriate fallbacks for all platforms.

### Is there a reason why this should be core and not an add-on in the asset library?

This is about improving the appearance of text when using system font aliases.

**Comment by lmurray** [comment]
It is worth pointing out that this proposal only seems to consider English and not any other language. Creating default font stacks requires dynamic selection of multiple fonts depending on what language packs are installed on the system. E.g. Japanese glyphs on Windows uses Yu Gothic UI in Windows 10+ or Meiryo UI in earlier versions instead of Segoe UI, but those fonts are only installed if the Japanese language pack is installed. Web browsers have lots of code for determining their default fonts, e.g. [Chromium's InitializeScriptFontMap()](https://chromium.googlesource.com/chromium/src.git/+/refs/heads/main/third_party/blink/renderer/platform/fonts/win/font_fallback_win.cc), and the linked Bootstrap font stack actually depends on this logic due to preferring the `system-ui` family over explicitly-defined fonts.

Useful references for Windows:

- [Script and font support in Windows](https://learn.microsoft.com/en-us/globalization/input/font-support)
- [Windows 11 font list](https://learn.microsoft.com/en-us/typography/fonts/windows_11_font_list)

**Comment by bruvzg** [comment]
> It is worth pointing out that this proposal only seems to consider English and not any other language.

Indeed, simply selecting `Segoe UI` will only give support for a limited set of languages (Godot will load only a single font file with the matching name). And since we are not using system text rendering APIs, we can't rely on its fallback system and will need to have fallback font lists for each platform.

In case of macOS and Linux, it's probably `Noto` fonts.

`SystemFont` can have multiple font names specified (to pick the first available one), and can have a list of fallback, so we probably should add a subclass (a `SystemUIFont` ?) with preconfigured lists.

**Comment by lmurray** [comment]
While outside the scope of this proposal, it'll be worth considering at some point in the future that some users will have a large amount of default fonts available on their system but not all are actually required to be loaded by Godot as the glyphs are never rendered. I'm under the impression that Chromium only loads the minimal amount of default fonts required based on what glyphs are attempted to be rendered and it does asynchronous loading of extra fonts once new glyphs are detected. Without this lazy loading of fonts Godot would have to load potentially 30+ system fonts (based on Chromium's font list above) at launch with most of them never used.

**Comment by Calinou** [comment]
> While outside the scope of this proposal, it'll be worth considering at some point in the future that some users will have a large amount of default fonts available on their system but not all are actually required to be loaded by Godot as the glyphs are never rendered. I'm under the impression that Chromium only loads the minimal amount of default fonts required based on what glyphs are attempted to be rendered and it does asynchronous loading of extra fonts once new glyphs are detected. Without this lazy loading of fonts Godot would have to load potentially 30+ system fonts (based on Chromium's font list above) at launch with most of them never used.

This is something I mentioned in https://github.com/godotengine/godot-proposals/issues/5779#issuecomment-1320983946.

**Comment by bruvzg** [comment]
Even loading all system font takes only a second - https://github.com/godotengine/godot/pull/56558#issuecomment-1207479137, so loading only a minimal set of fonts should not be too bad.

**Comment by Zireael07** [comment]
> stuttering the first time a glyph from a font is needed.

How big a stutter are we talking about?

**Comment by bruvzg** [comment]
> How big a stutter are we talking about?

It depends on a font, but since it will happen only once per language/script, not for every missing glyph, it's probably won't be an issue.

> And since we are not using system text rendering APIs, we can't rely on its fallback system and will need to have fallback font lists for each platform.

It seems to be possible to actually request the system to pick a font for a specific text -  https://github.com/godotengine/godot/pull/68995. Loading fonts seems to be working fine (tested on Linux, macOS and Windows, with a few simple cases), but performance is not tested yet.
